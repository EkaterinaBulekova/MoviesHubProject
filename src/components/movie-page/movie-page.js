import React from "react";
import { Header } from "../header/header";
import MovieDetail from "../movie-detail/movie-detail";
import Button from "../button/button";
import { ErrorBoundary } from "../error-boundary/error-boundary";
import FilmsByGenre from '../films-by-genre/films-by-genre';
import MovieList from "../movie-list/movie-list";

export class MoviePage extends React.Component{
  constructor(props) {
    super(props);
    this.state = {
      movie: this.props.movie,
      movies: []
    };
  }

  getFilter(genres){
    return {
      searchBy: "genres",
      search: genres,
      sortBy: "release_date"};
  }

  componentDidMount(){
    this.props.getData(this.getFilter(this.state.movie.genres))
    .then(result => this.setState({movies: result.data}))
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (nextState.movie.id !== this.state.movie.id){
      this.props.getData(this.getFilter(nextState.movie.genres))
      .then(result => this.setState({movies: result.data}))
      return false;
    }
    return true;
  }

  onMovieClick = (movie) =>()=>{
    this.setState({movie: movie});
  }

  render(){
    const {movie, movies} = this.state;
    console.log('render '+ movie.genres);
    return (
      <div className="page">
        <Header name={this.props.name}>
          <Button className="search-button" name="SEARCH" onClick={this.props.onReturn}/>
          <MovieDetail movie={movie}/>
          <FilmsByGenre genres={movie.genres}/>
        </Header>
        <ErrorBoundary className="global-error">
          <MovieList movies={movies.filter(curmovie => curmovie.id !== movie.id)} onClick={this.onMovieClick}/>
        </ErrorBoundary>
      </div>
    );
  }
}

