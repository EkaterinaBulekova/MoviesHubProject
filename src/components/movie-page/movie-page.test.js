import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Promise } from 'q';
import { MoviePage } from './movie-page';

describe('<MoviePage />', () => {
  it('should render as expect', () => {
    const value = {data:
      [
      {id:1, title:'movie1', genres:'testGenre'},
      {id:2, title:'movie2', genres:'testGenre'},
      {id:3, title:'movie3', genres:'testGenre'}
    ]}
    const testOnClick = jest.fn(); 
    const testGetData = jest.fn().mockImplementation(() => Promise.resolve(value)); 
    const testName = 'test'; 
    const component = mount(<MoviePage name={testName} getData={testGetData} movie={value.data[0]} onReturn={testOnClick}/>);
    const testMovieClick = jest.fn();
    component.update();

    expect(toJson(component)).toMatchSnapshot();

    expect(component.find('.films-by-genre-text').text()).toBe("Films by "+value.data[0].genres+" genre");

    component.find('button').simulate('click');
    expect(testOnClick).toHaveBeenCalled();

    return component.instance().getMoviesByGenre({genres:'genre'})
    .then(()=>expect(component.state().movies).toHaveLength(3))
    .then(()=>component.unmount());
  });
});