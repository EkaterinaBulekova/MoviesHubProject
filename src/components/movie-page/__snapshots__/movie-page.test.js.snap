// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MoviePage /> should render as expect 1`] = `
<MoviePage
  getData={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "search": "testGenre",
            "searchBy": "genres",
            "sortBy": "release_date",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {
            "inspect": [Function],
            "promiseDispatch": [Function],
            "valueOf": [Function],
          },
        },
      ],
    }
  }
  movie={
    Object {
      "genres": "testGenre",
      "id": 1,
      "title": "movie1",
    }
  }
  name="test"
  onReturn={[MockFunction]}
>
  <div
    className="page"
  >
    <Header
      name="test"
    >
      <div
        className="app-header"
      >
        <div
          className="header-name"
        >
          test
        </div>
        <Button
          className="search-button"
          name="SEARCH"
          onClick={[MockFunction]}
        >
          <button
            className="btn search-button"
            name="SEARCH"
            onClick={[MockFunction]}
            type="button"
          >
            SEARCH
          </button>
        </Button>
        <MovieDetail
          movie={
            Object {
              "genres": "testGenre",
              "id": 1,
              "title": "movie1",
            }
          }
        >
          <div
            className="movie-detail"
          >
            <div
              className="movie-detail-img-box"
            >
              <img
                className="movie-detail-img"
              />
            </div>
            <div
              className="movie-detail-info"
            >
              <div
                className="movie-detail-group"
              >
                <div
                  className="movie-detail-title"
                >
                  movie1
                </div>
                <div
                  className="movie-detail-rating"
                />
              </div>
              <div
                className="movie-detail-group"
              >
                <div
                  className="movie-detail-release-date"
                >
                  NaN
                </div>
                <div
                  className="movie-detail-duration"
                />
              </div>
              <div
                className="movie-detail-group"
              >
                <div
                  className="movie-detail-description"
                />
              </div>
            </div>
          </div>
        </MovieDetail>
        <FilmsByGenre
          genres="testGenre"
        >
          <div
            className="films-by-genre"
          >
            <div
              className="films-by-genre-text"
            >
              Films by testGenre genre
            </div>
          </div>
        </FilmsByGenre>
      </div>
    </Header>
    <ErrorBoundary
      className="global-error"
    >
      <MovieList
        movies={Array []}
        onClick={[Function]}
      >
        <ul
          className="movie-list-container"
        />
      </MovieList>
    </ErrorBoundary>
  </div>
</MoviePage>
`;
